File: Exercise/Very Hard/01. Cycle Detection/Code/Cycle Detection.cs
using System;
using System.Collections.Generic;

class Program {
    static List<int>[] graph;
    static bool[] visited;
    static bool[] recStack;

    static bool DFS(int node) {
        visited[node] = true;
        recStack[node] = true;

        foreach (int neighbor in graph[node]) {
            if (!visited[neighbor]) {
                if (DFS(neighbor)) {
                    return true;
                }
            } else if (recStack[neighbor]) {
                return true;
            }
        }

        recStack[node] = false;
        return false;
    }

    static bool HasCycle(int n, List<Tuple<int, int>> edges) {
        graph = new List<int>[n];
        visited = new bool[n];
        recStack = new bool[n];

        for (int i = 0; i < n; i++) {
            graph[i] = new List<int>();
        }

        foreach (var edge in edges) {
            graph[edge.Item1].Add(edge.Item2);
        }

        for (int node = 0; node < n; node++) {
            if (!visited[node]) {
                if (DFS(node)) {
                    return true;
                }
            }
        }

        return false;
    }

    static void Main() {
        int n = int.Parse(Console.ReadLine());
        int m = int.Parse(Console.ReadLine());
        var edges = new List<Tuple<int, int>>();
        for (int i = 0; i < m; i++) {
            var input = Console.ReadLine().Split();
            edges.Add(new Tuple<int, int>(int.Parse(input[0]), int.Parse(input[1])));
        }

        if (HasCycle(n, edges)) {
            Console.WriteLine("True");
        } else {
            Console.WriteLine("False");
        }
    }
}

File: Exercise/Very Hard/01. Cycle Detection/Code/Cycle Detection.py
def has_cycle(n, edges):
    from collections import defaultdict

    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)

    visited = [False] * n
    rec_stack = [False] * n

    def dfs(node):
        visited[node] = True
        rec_stack[node] = True

        for neighbor in graph[node]:
            if not visited[neighbor]:
                if dfs(neighbor):
                    return True
            elif rec_stack[neighbor]:
                return True

        rec_stack[node] = False
        return False

    for node in range(n):
        if not visited[node]:
            if dfs(node):
                return True

    return False

def main():
    n = int(input().strip())
    m = int(input().strip())
    edges = [tuple(map(int, input().strip().split())) for _ in range(m)]

    if has_cycle(n, edges):
        print("True")
    else:
        print("False")

if __name__ == "__main__":
    main()
